File: Graph/php/hasPath.php
<?php

declare(strict_types=1);

require_once __DIR__ . '/GraphUtils.php';

function hasPath(array $graph, string $src, string $dst): bool
{
    if ($src === $dst) {
        return true;
    }

    $visited = [];
    return hasPathDfs($graph, $src, $dst, $visited);
}

function hasPathDfs(array $graph, string $node, string $dst, array &$visited): bool
{
    if ($node === $dst) {
        return true;
    }
    if (isset($visited[$node])) {
        return false;
    }

    $visited[$node] = true;
    foreach ($graph[$node] ?? [] as $neighbor) {
        if (hasPathDfs($graph, $neighbor, $dst, $visited)) {
            return true;
        }
    }

    return false;
}

File: Graph/go/has_path.go
package graph

// HasPath reports whether dst is reachable from src in a directed graph.
func HasPath(g Graph, src, dst string) bool {
	if src == dst {
		return true
	}
	if _, ok := g[src]; !ok {
		return false
	}
	visited := map[string]struct{}{}
	return hasPathDFS(g, src, dst, visited)
}

func hasPathDFS(g Graph, node, dst string, visited map[string]struct{}) bool {
	if node == dst {
		return true
	}
	if _, seen := visited[node]; seen {
		return false
	}
	visited[node] = struct{}{}

	for _, neighbor := range g[node] {
		if hasPathDFS(g, neighbor, dst, visited) {
			return true
		}
	}
	return false
}
